src/rgw/rgw_sal_daos.cc

```diff
@@ -0,0 +1,2803 @@
+  }
+
+  bufferlist rbl;
+  rgw_bucket_dir_entry ent;
+  rbl.append(reinterpret_cast<char*>(value.data()), size);
+  auto iter = rbl.cbegin();
+  ent.decode(iter);
+
+  // Update object metadata
+  Attrs updateattrs = setattrs ? attrs : *setattrs;
+  if (delattrs) {
+    for (auto const& [attr, attrval] : *delattrs) {
+      updateattrs.erase(attr);
+    }
+  }
+
+  bufferlist wbl;
+  ent.encode(wbl);
+  encode(updateattrs, wbl);
+
+  // Write rgw_bucket_dir_entry into object xattr
+  ret = dfs_setxattr(get_daos_bucket()->dfs, dfs_obj, RGW_DIR_ENTRY_XATTR,
+                     wbl.c_str(), wbl.length(), 0);
+  if (ret != 0) {
+    ldpp_dout(dpp, 0) << "ERROR: failed to set xattr of daos object ("
+                      << get_bucket()->get_name() << ", " << get_key().to_str()
+                      << "): ret=" << ret << dendl;
+  }
+  return ret;
+}
+
+int DaosObject::get_obj_attrs(RGWObjectCtx* rctx, optional_yield y,
+                              const DoutPrefixProvider* dpp,
+                              rgw_obj* target_obj) {
+  ldpp_dout(dpp, 20) << "DEBUG: DaosObject::get_obj_attrs()" << dendl;
+  // TODO handle target_obj
+
+  // Get object's metadata (those stored in rgw_bucket_dir_entry)
+  int ret = open(dpp, DaosObjectOpen::Lookup);
+  vector<uint8_t> value(DFS_MAX_XATTR_LEN);
+  size_t size = value.size();
+  ret = dfs_getxattr(get_daos_bucket()->dfs, dfs_obj, RGW_DIR_ENTRY_XATTR,
+                     value.data(), &size);
+  if (ret != 0) {
+    ldpp_dout(dpp, 0) << "ERROR: failed to get xattr of daos object ("
+                      << get_bucket()->get_name() << ", " << get_key().to_str()
+                      << "): ret=" << ret << dendl;
+    return ret;
+  }
+
+  bufferlist bl;
+  rgw_bucket_dir_entry ent;
+  bl.append(reinterpret_cast<char*>(value.data()), size);
+  auto iter = bl.cbegin();
+  ent.decode(iter);
+  obj_size = ent.meta.size;
+  mtime = ent.meta.mtime;
+  decode(attrs, iter);
+  return ret;
+}
+
+int DaosObject::modify_obj_attrs(RGWObjectCtx* rctx, const char* attr_name,
+                                 bufferlist& attr_val, optional_yield y,
+                                 const DoutPrefixProvider* dpp) {
+  rgw_obj target = get_obj();
+  int r = get_obj_attrs(rctx, y, dpp, &target);
+  if (r < 0) {
+    return r;
+  }
+  set_atomic(rctx);
+  attrs[attr_name] = attr_val;
+  return set_obj_attrs(dpp, rctx, &attrs, nullptr, y, &target);
+}
+
+int DaosObject::delete_obj_attrs(const DoutPrefixProvider* dpp,
+                                 RGWObjectCtx* rctx, const char* attr_name,
+                                 optional_yield y) {
+  rgw_obj target = get_obj();
+  Attrs rmattr;
+  bufferlist bl;
+
+  set_atomic(rctx);
+  rmattr[attr_name] = bl;
+  return set_obj_attrs(dpp, rctx, nullptr, &rmattr, y, &target);
+}
+
+/* RGWObjectCtx will be moved out of sal */
+/* XXX: Placeholder. Should not be needed later after Dan's patch */
+void DaosObject::set_atomic(RGWObjectCtx* rctx) const { return; }
```

Comment:
These 3 routines are no more part of sal.

Reply:
I mean these comments no more hold true. RGWObjectCtx() is now made part of rados store and all these routines have default definition in sal.h to be used by the other stores.